### PR TITLE
block preprocess stage

### PR DESCRIPTION
**Pull request by marbar3778**
## Description

Opening this PR here to facilitate some discussion on a possible design. 

Currently, the approach taken is:

- When a proposer proposes the `txs` that would be included in a block are sent to the app. 
- The app does whatever they want and has the option to add a proof of their black magic in the response. 
- The returned proof and `txs` are added to the block. 

The data struct in the block now consists of a metadata field.

### CODE REVIEW

File: state/execution.go [hunk 1]
```diff
@@ -107,8 +107,28 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
...
+	metaData := processedBlockTxs.GetMetaData()
+
+	lp := len(ppt)
+	processedTxs := make(types.Txs, lp)
+	for i := 0; i < l; i++ {
+		processedTxs[i] = ppt[i]
+	}
 
-	return state.MakeBlock(height, txs, commit, evidence, proposerAddr)
+	return state.MakeBlock(height, processedTxs, metaData, commit, evidence, proposerAddr)
```
Comment (liamsi) [inline state/execution.go:131]: I think I understand now why you said this (merkelizing the shares into the NMT) can all happen in the app.

Comment (liamsi) [reply state/execution.go:131]: But I still think the actual computation of shares and writing into the NMT, should happen on the ll-core side.

File: state/execution.go [hunk 2]
```diff
@@ -112,14 +112,30 @@ func (blockExec *BlockExecutor) CreateProposalBlock(
...
-	//      https://github.com/lazyledger/lazyledger-specs/blob/53e5f350838f1e0785ad670704bf91dac2f4f5a3/specs/block_proposer.md#deciding-on-a-block-size
-	//      Here, we instead assume a fixed (max) square size instead.
-	//  2. feed them into MakeBlock below:
-	return state.MakeBlock(height, txs, evidence, nil, nil, commit, proposerAddr)
+	l := len(txs)
+	bzs := make([][]byte, l)
+	for i := 0; i < l; i++ {
+		bzs[i] = txs[i]
+	}
+
```
Comment (marbar3778) [inline state/execution.go:120]:
```suggestion
// TODO(ismail):
	//  1. get those intermediate state roots & messages either from the
	//     mempool or from the abci-app
	//  1.1 at this point we should now the square / block size:
	//      https://github.com/lazyledger/lazyledger-specs/blob/53e5f350838f1e0785ad670704bf91dac2f4f5a3/specs/block_proposer.md#deciding-on-a-block-size
	//      Here, we instead assume a fixed (max) square size instead.
	//  2. feed them into MakeBlock below:
```

File: state/validation_test.go [hunk 3]
```diff
@@ -84,7 +84,7 @@ func TestValidateBlockHeader(t *testing.T) {
 			Invalid blocks don't pass
 		*/
 		for _, tc := range testCases {
-			block, _ := state.MakeBlock(height, makeTxs(height), nil, nil, nil, lastCommit, proposerAddr)
+			block, _ := state.MakeBlock(height, makeTxs(height), []byte("proof"), nil, nil, nil, lastCommit, proposerAddr)
```
Comment (marbar3778) [inline state/validation_test.go:87]:
```suggestion
			block, _ := state.MakeBlock(height, makeTxs(height), nil, nil, nil, nil, lastCommit, proposerAddr)
```

File: store/store_test.go [hunk 4]
```diff
@@ -50,7 +50,7 @@ func makeTxs(height int64) (txs []types.Tx) {
 }
 
 func makeBlock(height int64, state sm.State, lastCommit *types.Commit) *types.Block {
-	block, _ := state.MakeBlock(height, makeTxs(height), nil, nil, nil, lastCommit, state.Validators.GetProposer().Address)
+	block, _ := state.MakeBlock(height, makeTxs(height), []byte("meta_data"), nil, nil, nil, lastCommit, state.Validators.GetProposer().Address)
```
Comment (marbar3778) [inline store/store_test.go:53]:
```suggestion
	block, _ := state.MakeBlock(height, makeTxs(height), nil, nil, nil, nil, lastCommit, state.Validators.GetProposer().Address)
```

File: types/block_test.go [hunk 5]
```diff
@@ -117,13 +117,13 @@ func TestBlockValidateBasic(t *testing.T) {
 
 func TestBlockHash(t *testing.T) {
 	assert.Nil(t, (*Block)(nil).Hash())
-	assert.Nil(t, MakeBlock(int64(3), []Tx{Tx("Hello World")}, nil, nil, nil, nil).Hash())
+	assert.Nil(t, MakeBlock(int64(3), []Tx{Tx("Hello World")}, []byte("proof"), nil, nil, nil, nil).Hash())
```
Comment (marbar3778) [inline types/block_test.go:120]:
```suggestion
	assert.Nil(t, MakeBlock(int64(3), []Tx{Tx("Hello World")}, nil, nil, nil, nil, nil).Hash())
```

File: types/block_test.go [hunk 6]
```diff
@@ -117,13 +117,13 @@ func TestBlockValidateBasic(t *testing.T) {
...
 }
 
 func TestBlockMakePartSet(t *testing.T) {
 	assert.Nil(t, (*Block)(nil).MakePartSet(2))
 
-	partSet := MakeBlock(int64(3), []Tx{Tx("Hello World")}, nil, nil, nil, nil).MakePartSet(1024)
+	partSet := MakeBlock(int64(3), []Tx{Tx("Hello World")}, []byte("proof"), nil, nil, nil, nil).MakePartSet(1024)
```
Comment (marbar3778) [inline types/block_test.go:126]:
```suggestion
	partSet := MakeBlock(int64(3), []Tx{Tx("Hello World")}, nil, nil, nil, nil, nil).MakePartSet(1024)
```

File: types/block_test.go [hunk 7]
```diff
@@ -141,7 +141,7 @@ func TestBlockMakePartSetWithEvidence(t *testing.T) {
 	ev := NewMockDuplicateVoteEvidenceWithValidator(h, time.Now(), vals[0], "block-test-chain")
 	evList := []Evidence{ev}
 
-	partSet := MakeBlock(h, []Tx{Tx("Hello World")}, evList, nil, nil, commit).MakePartSet(512)
+	partSet := MakeBlock(h, []Tx{Tx("Hello World")}, []byte("proof"), evList, nil, nil, commit).MakePartSet(512)
```
Comment (marbar3778) [inline types/block_test.go:144]:
```suggestion
	partSet := MakeBlock(h, []Tx{Tx("Hello World")}, nil, evList, nil, nil, commit).MakePartSet(512)
```

File: types/block_test.go [hunk 8]
```diff
@@ -158,15 +158,15 @@ func TestBlockHashesTo(t *testing.T) {
 	ev := NewMockDuplicateVoteEvidenceWithValidator(h, time.Now(), vals[0], "block-test-chain")
 	evList := []Evidence{ev}
 
-	block := MakeBlock(h, []Tx{Tx("Hello World")}, evList, nil, nil, commit)
+	block := MakeBlock(h, []Tx{Tx("Hello World")}, []byte("proof"), evList, nil, nil, commit)
```
Comment (marbar3778) [inline types/block_test.go:161]:
```suggestion
	block := MakeBlock(h, []Tx{Tx("Hello World")}, nil, evList, nil, nil, commit)
```

File: types/block_test.go [hunk 9]
```diff
@@ -158,15 +158,15 @@ func TestBlockHashesTo(t *testing.T) {
...
 	block.ValidatorsHash = valSet.Hash()
 	assert.False(t, block.HashesTo([]byte{}))
 	assert.False(t, block.HashesTo([]byte("something else")))
 	assert.True(t, block.HashesTo(block.Hash()))
 }
 
 func TestBlockSize(t *testing.T) {
-	size := MakeBlock(int64(3), []Tx{Tx("Hello World")}, nil, nil, nil, nil).Size()
+	size := MakeBlock(int64(3), []Tx{Tx("Hello World")}, []byte("proof"), nil, nil, nil, nil).Size()
```
Comment (marbar3778) [inline types/block_test.go:169]:
```suggestion
	size := MakeBlock(int64(3), []Tx{Tx("Hello World")}, nil, nil, nil, nil, nil).Size()
```

File: types/block_test.go [hunk 10]
```diff
@@ -177,7 +177,7 @@ func TestBlockString(t *testing.T) {
 	assert.Equal(t, "nil-Block", (*Block)(nil).StringIndented(""))
 	assert.Equal(t, "nil-Block", (*Block)(nil).StringShort())
 
-	block := MakeBlock(int64(3), []Tx{Tx("Hello World")}, nil, nil, nil, nil)
+	block := MakeBlock(int64(3), []Tx{Tx("Hello World")}, []byte("proof"), nil, nil, nil, nil)
```
Comment (marbar3778) [inline types/block_test.go:180]:
```suggestion
	block := MakeBlock(int64(3), []Tx{Tx("Hello World")}, nil, nil, nil, nil, nil)
```

File: types/block_test.go [hunk 11]
```diff
@@ -649,17 +649,17 @@ func TestBlockIDValidateBasic(t *testing.T) {
 func TestBlockProtoBuf(t *testing.T) {
 	h := tmrand.Int63()
 	c1 := randCommit(time.Now())
-	b1 := MakeBlock(h, []Tx{Tx([]byte{1})}, []Evidence{}, nil, nil, &Commit{Signatures: []CommitSig{}})
+	b1 := MakeBlock(h, []Tx{Tx([]byte{1})}, []byte("proof"), []Evidence{}, nil, nil, &Commit{Signatures: []CommitSig{}})
```
Comment (marbar3778) [inline types/block_test.go:652]:
```suggestion
	b1 := MakeBlock(h, []Tx{Tx([]byte{1})}, nil, []Evidence{}, nil, nil, &Commit{Signatures: []CommitSig{}})
```

File: types/block_test.go [hunk 12]
```diff
@@ -649,17 +649,17 @@ func TestBlockIDValidateBasic(t *testing.T) {
...
 	b1.ProposerAddress = tmrand.Bytes(crypto.AddressSize)
 
-	b2 := MakeBlock(h, []Tx{Tx([]byte{1})}, []Evidence{}, nil, nil, c1)
+	b2 := MakeBlock(h, []Tx{Tx([]byte{1})}, []byte("proof"), []Evidence{}, nil, nil, c1)
```
Comment (marbar3778) [inline types/block_test.go:655]:
```suggestion
	b2 := MakeBlock(h, []Tx{Tx([]byte{1})}, nil, []Evidence{}, nil, nil, c1)
```

File: types/event_bus_test.go [hunk 13]
```diff
@@ -75,7 +75,7 @@ func TestEventBusPublishEventNewBlock(t *testing.T) {
 		}
 	})
 
-	block := MakeBlock(0, []Tx{}, []Evidence{}, nil, nil, nil)
+	block := MakeBlock(0, []Tx{}, []byte("proof"), []Evidence{}, nil, nil, nil)
```
Comment (marbar3778) [inline types/event_bus_test.go:78]:
```suggestion
	block := MakeBlock(0, []Tx{}, nil, []Evidence{}, nil, nil, nil)
```

File: types/event_bus_test.go [hunk 14]
```diff
@@ -234,7 +234,7 @@ func TestEventBusPublishEventNewBlockHeader(t *testing.T) {
...
```
Comment (marbar3778) [inline types/event_bus_test.go:237]:
```suggestion
	block := MakeBlock(0, []Tx{}, nil, []Evidence{}, nil, nil, nil)
```

**Pull request closed without merging**